File: rawmatch.java
package com.contentguard.model;
import java.util.Objects;
public class RawMatch {
    private String submissionId;
    private String matchedSourceId;
    private double exactScore;
    private double structuralScore;
    private String matchedFragment;
    private int matchStartIndex;
    private int matchEndIndex;
    public RawMatch() {}
    public RawMatch(String submissionId,
                    String matchedSourceId,
                    double exactScore,
                    double structuralScore,
                    String matchedFragment,
                    int matchStartIndex,
                    int matchEndIndex) {
        this.submissionId = Objects.requireNonNull(submissionId);
        this.matchedSourceId = Objects.requireNonNull(matchedSourceId);
        setExactScore(exactScore);
        setStructuralScore(structuralScore);
        this.matchedFragment = matchedFragment;
        this.matchStartIndex = matchStartIndex;
        this.matchEndIndex = matchEndIndex;
    }
    public String getSubmissionId() {
        return submissionId;
    }
    public void setSubmissionId(String submissionId) {
        this.submissionId = submissionId;
    }
    public String getMatchedSourceId() {
        return matchedSourceId;
    }
    public void setMatchedSourceId(String matchedSourceId) {
        this.matchedSourceId = matchedSourceId;
    }
    public double getExactScore() {
        return exactScore;
    }
    public void setExactScore(double exactScore) {
        validateScore(exactScore, "Exact score");
        this.exactScore = exactScore;
    }
    public double getStructuralScore() {
        return structuralScore;
    }
    public void setStructuralScore(double structuralScore) {
        validateScore(structuralScore, "Structural score");
        this.structuralScore = structuralScore;
    }
    public String getMatchedFragment() {
        return matchedFragment;
    }
    public void setMatchedFragment(String matchedFragment) {
        this.matchedFragment = matchedFragment;
    }
    public int getMatchStartIndex() {
        return matchStartIndex;
    }
    public void setMatchStartIndex(int matchStartIndex) {
        this.matchStartIndex = matchStartIndex;
    }
    public int getMatchEndIndex() {
        return matchEndIndex;
    }
    public void setMatchEndIndex(int matchEndIndex) {
        this.matchEndIndex = matchEndIndex;
    }
    private void validateScore(double value, String field) {
        if (value < 0.0 || value > 1.0) {
            throw new IllegalArgumentException(field + " must be between 0 and 1");
        }
    }
    @Override
    public String toString() {
        return String.format(
                "RawMatch[submission=%s, source=%s, exact=%.2f, structural=%.2f]",
                submissionId,
                matchedSourceId,
                exactScore,
                structuralScore
        );
    }
}

File: ranking/rawmatch.cpp
#ifndef RAWMATCH_HPP
#define RAWMATCH_HPP
#include <string>
#include <stdexcept>
#include <string>
#include <stdexcept>
#include <iostream>

using namespace std;
class RawMatch {
private:
    string submissionId;
    string matchedSourceId;
    double exactScore;
    double structuralScore;
    string matchedFragment;
    int matchStartIndex;
    int matchEndIndex;
    void validateScore(double value) {
        if (value < 0.0 || value > 1.0) {
            throw invalid_argument("Score must be between 0 and 1");
        }
    }
public:
    RawMatch(
        const string& submissionId,
        const string& matchedSourceId,
        double exactScore,
        double structuralScore,
        const string& matchedFragment,
        int matchStartIndex,
        int matchEndIndex
    ) : submissionId(submissionId),
        matchedSourceId(matchedSourceId),
        matchedFragment(matchedFragment),
        matchStartIndex(matchStartIndex),
        matchEndIndex(matchEndIndex) {
        validateScore(exactScore);
        validateScore(structuralScore);
        this->exactScore = exactScore;
        this->structuralScore = structuralScore;
    }
    string getSubmissionId() const {
        return submissionId;
    }
    string getMatchedSourceId() const {
        return matchedSourceId;
    }
    double getExactScore() const {
        return exactScore;
    }
    double getStructuralScore() const {
        return structuralScore;
    }
    string getMatchedFragment() const {
        return matchedFragment;
    }
    int getMatchStartIndex() const {
        return matchStartIndex;
    }
    int getMatchEndIndex() const {
        return matchEndIndex;
    }
};
#endif
